gui/widgets/grid.cpp

```diff
@@ -0,0 +1,640 @@
+		_grid->_tray->runModal();
+	}
+}
+
+#pragma mark -
+
+GridItemTray::GridItemTray(GuiObject *boss, int x, int y, int w, int h, int entryID, GridWidget *grid)
+	: Dialog(x, y, w, h), CommandSender(boss) {
+		
+	_entryID = entryID;
+	_boss = boss;
+	_grid = grid;
+	
+	int buttonWidth = w / 3;
+	int buttonHeight = h / 3;
+
+	PicButtonWidget *playButton = new PicButtonWidget(this, (buttonWidth / 3), buttonHeight / 3, (buttonWidth / 3) + buttonWidth * 2, buttonHeight, U32String("Play"), kPlayButtonCmd);
+	PicButtonWidget *loadButton = new PicButtonWidget(this, (buttonWidth / 3), (buttonHeight * 5) / 3, buttonWidth, buttonHeight, U32String("Saves"), kLoadButtonCmd);
+	PicButtonWidget *editButton = new PicButtonWidget(this, buttonWidth + 2 * (buttonWidth / 3), (buttonHeight * 5) / 3, buttonWidth, buttonHeight, U32String("Edit"), kEditButtonCmd);
+	
+	playButton->useThemeTransparency(true);
+	loadButton->useThemeTransparency(true);
+	editButton->useThemeTransparency(true);
+
+	playButton->setGfxFromTheme("button_play.bmp", 0, false);
+	loadButton->setGfxFromTheme("button_load.bmp", 0, false);
+	editButton->setGfxFromTheme("button_options.bmp", 0, false);
+}
+
+void GridItemTray::handleCommand(CommandSender *sender, uint32 cmd, uint32 data) {
+	switch (cmd) {
+	case kPlayButtonCmd:
+		close();
+		sendCommand(kPlayButtonCmd, _entryID);
+		break;
+	case kLoadButtonCmd:
+		close();
+		sendCommand(kLoadButtonCmd, _entryID);
+		break;
+	case kEditButtonCmd:
+		close();
+		sendCommand(kEditButtonCmd, _entryID);
+		break;
+	case kCloseCmd:
+		close();
+		break;
+	default:
+		Dialog::handleCommand(sender, cmd, 0);
+		break;
+	}
+}
+
+void GridItemTray::handleMouseDown(int x, int y, int button, int clickCount) {
+	Dialog::handleMouseDown(x, y, button, clickCount);
+	if ((x < 0 || x > _w) || (y > _h || y < -(_grid->_gridItemHeight))) {
+		// Close on clicking outside
+		close();
+	} else if (y < 0 && clickCount >= 2) {
+		// Run on double clicking thumbnail
+		close();
+		sendCommand(kItemDoubleClickedCmd, _entryID);
+	}
+}
+
+void GridItemTray::handleMouseWheel(int x, int y, int direction) {
+	Dialog::handleMouseWheel(x, y, direction);
+	close();
+}
+
+void GridItemTray::handleMouseMoved(int x, int y, int button) {
+	Dialog::handleMouseMoved(x, y, button);
+	if ((x < 0 || x > _w) || (y > _h || y < -(_grid->_gridItemHeight))) {
+		// Close on going outside
+		close();
+	}
+}
+
+#pragma mark -
+
+Graphics::ManagedSurface *loadSurfaceFromFile(const Common::String &name) {
+	Graphics::ManagedSurface *surf = nullptr;
+	const Graphics::Surface *srcSurface = nullptr;
+	const String path = String::format("%s/%s", ConfMan.get("iconpath").c_str(), name.c_str());
+	if (name.hasSuffix(".png")) {
+#ifdef USE_PNG
+		Image::PNGDecoder decoder;
+		Common::FSNode fileNode(path);
+		Common::SeekableReadStream *stream = fileNode.createReadStream();
+		if (stream) {
+			if (!decoder.loadStream(*stream))
+				warning("Error decoding PNG");
```

Comment:
Is it safe to continue in this case?

Reply:
I am not sure, but it would be safer in this case to just return a nullptr, which I have done in commit 1a5695f.